### ISSUE TITLE
Add support for this to be an additional OVMF instead of a permanent and global one

### ISSUE THREAD

**Issue by tiagofreire-pt** [thenickdude/pve-edk2-firmware#1]
Like the subject says, it can be preferable sometimes to make this an alternative option instead of a permanent one.

Thanks in advance.

**Comment by autonomy-rocks** [comment]
1. Extract the .deb archive using `ar`. You will get a `data.tar.xz` archive. Extract the contents of this.
2. Copy the contents of `data.tar.xz/usr/share/pve-edk2-firmware` to `/opt/pve-edk2-firmware-macos` (or your preferred directory) on the Proxmox host.
3. In the `args:` section of `/etc/pve/qemu-server/YOUR-VM-ID-HERE.conf` add the following at the end of the commands you added from the tutorial:
  `-drive 'if=pflash,unit=0,format=raw,readonly,file=/opt/pve-edk2-firmware-macos//OVMF_CODE.fd' -drive 'if=pflash,unit=1,format=raw,id=drive-efidisk0,file=/dev/pve/vm-100-disk-1'`
4. Go to the `Hardware->BIOS` settings for your VM and select `Default (SeaBIOS)`.

This worked for me without modifying the OVMF used for every VM on the server.

**Comment by autonomy-rocks** [comment]
I've created a separate package that doesn't conflict with the OVMF installed by Proxmox. So you can use similar steps to the above to use a patched OVMF. The package I created also resolves the graphical glitch on cold-boot issue caused by the screen resolution not taking.

https://gitlab.com/pve-edk2-firmware-clover/pve-edk2-firmware-clover/blob/clover-patch/README.md

**Comment by victorhooi** [comment]
@tstivers1990 Are there any disadvantages to your patched version?

And the hardcoded resolution to 1920x1080 only affects the BIOS screen, correct - it doesn't affect the actual MacOS environment?

**Comment by autonomy-rocks** [comment]
> @tstivers1990 Are there any disadvantages to your patched version?

Not that I've found. Works the same, but you don't have to replace the built in file, which means any other VMs that use that file won't be affected.

> And the hardcoded resolution to 1920x1080 only affects the BIOS screen, correct - it doesn't affect the actual MacOS environment?

It will indeed affect the actual MacOS environment. If you want a different resolution, you will need a custom build with that resolution. You should be able to fork my project and patch it for your preferred resolution. I'm open to improving this in my fork. But I can't think of a maintainable or straight forward way to go about it.

**Comment by victorhooi** [comment]
Yeah, the workarounds/hoops are fine - I think the main issue will be that you can no longer change the resolution in MacOS.

So if we have a team of people testing it, they may have different laptops/desktops/monitors etc. - and want to change the resolution to suit.

I didn't realise changing the resolution in the BIOS like that would also limit changing the resolution in the final machine as well.

**Comment by autonomy-rocks** [comment]
> Yeah, the workarounds/hoops are fine - I think the main issue will be that you can no longer change the resolution in MacOS.
> 
> So if we have a team of people testing it, they may have different laptops/desktops/monitors etc. - and want to change the resolution to suit.

Yeah. I don't know how prevalent a scenario that is, though. At least, in my case, I don't care much about the resolution as I don't actually interact with the OS all that much. It just acts as a server for things that can't be done with Windows/Linux, like building iOS apps.

One potential option is to automate the patches and build a package for each of the standard resolutions. I imagine that's reasonably feasible to accomplish. But it's probably something better discussed in an issue in that project, rather than this issue.

**Comment by victorhooi** [comment]
@tstivers1990 I'm trying to follow your guide at https://gitlab.com/pve-edk2-firmware-clover/pve-edk2-firmware-clover/blob/clover-patch/README.md

I was able to install your .DEB file from the releases page.

However, I am getting an error when starting the VM:

```
Logs
kvm: -drive if=pflash,unit=1,format=raw,id=drive-efidisk0,file=/dev/pve/vm-540-disk-1: Could not open '/dev/pve/vm-540-disk-1': No such file or directory
TASK ERROR: start failed: QEMU exited with code 1
```

Your guide seems to assume you're using LVM, right? (I don't have a /dev/pve director - which I think only exists if you're using LVM).

My Proxmox cluster uses ZFS for the root and Ceph for the actual VM disk images.

Any idea on how to update it for that sort of setup?

https://gitlab.com/pve-edk2-firmware-clover/pve-edk2-firmware-clover/-/issues/2
